### PR TITLE
[spl-record] Remove borsh dependency from spl-record program

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
Currently, the zero-knowledge proofs that are required for the token-2022 program barely fits in a single transaction. The spl-record program provides a convenient way to break up the proofs into byte chunks to be submitted on chain. However, the record program depends on borsh serialization, which creates issues when combined with the zk-token-proof program.

#### Summary of changes
Remove the borsh dependency from the record program and use bytemuck for the serialization of data.

### CODE REVIEW

**Review by joncinque — APPROVED** [review]
Everything looks great! No show-stopping concerns, just some suggestions to make this better in subsequent work, if you'd like

File: record/program/src/instruction.rs [hunk 1]
```diff
@@ -2,15 +2,16 @@
...
+        program_error::ProgramError,
         pubkey::Pubkey,
     },
+    std::mem::size_of,
 };
 
 /// Instructions supported by the program
-#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, PartialEq)]
+#[derive(Clone, Debug, PartialEq)]
 pub enum RecordInstruction {
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply record/program/src/instruction.rs:15]: Yes, this is a good idea. Updated!

File: record/program/src/state.rs [hunk 2]
```diff
@@ -17,11 +18,15 @@ pub struct RecordData {
     pub data: Data,
 }
 
+/// The length of the data contained in the account for testing.
+const DATA_SIZE: usize = 8;
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply record/program/src/state.rs:22]: Okay, I think I can probably do this in a follow-up PR.

File: record/program/src/instruction.rs [hunk 3]
```diff
@@ -53,28 +54,70 @@ pub enum RecordInstruction {
...
+
+    /// Packs a [RecordInstruction] into a byte buffer.
+    pub fn pack(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(size_of::<Self>());
+        match self {
+            Self::Initialize => buf.push(0),
+            Self::Write { offset, data } => {
+                buf.push(1);
+                buf.extend_from_slice(&offset.to_le_bytes());
+                buf.extend_from_slice(data);
```
[the remark this answers is not in the document]
Comment (samkim-crypto) [reply record/program/src/instruction.rs:100]: Yes, good point. I added the vector length in the serialization.

**Review by joncinque — APPROVED** [review]
Looks great!

**Pull request merged**